scripts/migrations/2020-10-miro-migration/transfer_packager.py

```diff
@@ -0,0 +1,195 @@
+
+    assert local_content_length == expected_content_length, (
+        "Content length mismatch "
+        f"({local_content_length} != {expected_content_length}): "
+        f"{file_location}"
+    )
+
+
+def _download_s3_object(s3_client, s3_bucket, s3_key, target_folder, prefix):
+    assert s3_key.startswith(prefix), f"{s3_key} does not start with {prefix}"
+
+    file_name = s3_key[len(prefix) :].strip("/")
+    file_location = os.path.join(target_folder, file_name)
+
+    s3_content_length = get_s3_content_length(
+        s3_client=s3_client, s3_bucket=s3_bucket, s3_key=s3_key
+    )
+
+    def _download():
+        required_dir = os.path.dirname(file_location)
+        os.makedirs(required_dir, exist_ok=True)
+        s3_client.download_file(Bucket=s3_bucket, Key=s3_key, Filename=file_location)
+
+    # Check if we already have this file
+    if os.path.isfile(file_location):
+        # If the length doesn't match overwrite
+        if not os.path.getsize(file_location) == s3_content_length:
+            _download()
+    else:
+        _download()
+
+    _check_local_content_length(
+        file_location=file_location, expected_content_length=s3_content_length
+    )
+
+
+def _download_objects_from_s3(s3_client, target_folder, s3_bucket, s3_key_list, prefix):
+    def _get_s3_object(s3_key):
+        _download_s3_object(
+            s3_client=s3_client,
+            s3_bucket=s3_bucket,
+            s3_key=s3_key,
+            target_folder=target_folder,
+            prefix=prefix,
+        )
+
+    # itertools below expects an iterator
+    s3_key_list_iter = iter(s3_key_list)
+
+    with tqdm(total=len(s3_key_list)) as pbar:
+        # From https://alexwlchan.net/2019/10/adventures-with-concurrent-futures/
+        with concurrent.futures.ThreadPoolExecutor() as executor:
+            # Schedule the first N futures.  We don't want to schedule them all
+            # at once, to avoid consuming excessive amounts of memory.
+            futures = {
+                executor.submit(_get_s3_object, task): task
+                for task in itertools.islice(s3_key_list_iter, S3_DOWNLOAD_CONCURRENCY)
+            }
+
+            while futures:
+                # Wait for the next future to complete.
+                done, _ = concurrent.futures.wait(
+                    futures, return_when=concurrent.futures.FIRST_COMPLETED
+                )
+
+                for fut in done:
+                    futures.pop(fut)
+
+                pbar.update(len(done))
+
+                # Schedule the next set of futures.  We don't want more than N futures
+                # in the pool at a time, to keep memory consumption down.
+                for task in itertools.islice(s3_key_list_iter, len(done)):
+                    fut = executor.submit(_get_s3_object, task)
+                    futures[fut] = task
+
+        target_folder_count = sum(
+            [len(files) for r, d, files in os.walk(target_folder)]
```

Comment:
nit: Python convention is to use `_` for unused variables, i.e.
```suggestion
            [len(files) for _, _, files in os.walk(target_folder)]
```